### DISCUSSION TITLE
Sinowealth BMS on USB Hub, dedicate Serial Port?

### DISCUSSION THREAD

**Discussion by LundSoftwares** [Louisvdw/dbus-serialbattery#692] [General]
My (BMS related) setup is as follows:
- VenusOS on RPi4, placed in mid part of boat
- 5m USB extender cable back to battery compartment
- Active powered USB hub with following devices:
    - Daly/Sinowealth BMS to 4cell LiFePO4
    - Victron Smart Shunt

My problem is that when I do a complete reboot of the VenusOS system, I often end up with two devices: the correct Sinowealth BMS + another Sinowealth device with faulty values (25v total voltage, -2'000'000A current, 150% SoC and so on). The SmartShunt is never found so I assume that the "faulty" Sinowealth is actually the SmartShunt that's decoded wrongly? 
(Edit, did another search in Discussions and found that SmartShunt has been identified as Sinowealth earlier, but the solution there was to disable Sinowealth from config, but thats not an option in my case)
The only way that seems to work (atleast yesterday) is if I unplug the BMS, restart the RPi and let it find the SmartShunt, then plug in the BMS again and it will be found.

I guess the root cause here is how RPi handles Serial ports and the USB Hub, but since it works perfectly when I get both devices identified there should be possible to find a solution?
One thing that I would like to try is to hardcode the serial port where the BMS is located and see if that helps. I'm not that into Python, but is there somewhere (I would guess row 111 in dbus-serialbattery.py: port = /dev/ttyUSB2 ?) that I could change so only a dedicated port is scanned and not the entire array of ports?
Need to update to v1.0 first ofcourse...

Thanks for an awesome Addon to VenusOS!

![image](https://github.com/Louisvdw/dbus-serialbattery/assets/23386303/a2e63e1c-4018-473c-9aec-af2f8d692497)

**Comment by transistorgit** [comment]
Using fixed physical ports for devices would need configuration of udev rules. It is feasable, but cumbersome (I did that on a commerical project). I just wanted to give you a hint into that direction, but maybe others here know sinowealth and have a better suggestion.

(personally I would propose to improve the sinowealth driver so that it won't infere with the smartshunt)

**Reply by transistorgit** [reply]
I just had a look into the sinowealth driver. there is plenty of room for improvement. It literally takes anything that comes in, pulls out some bits and numbers and presents these as measurements. There is hardly any check and especially no checksum evaluation. So not surprising that it detects some other device as sino bms (ANT BMS driver has the same problem and is now also disabled by default).

I also checked the internet shortly, there is no information to be found (easily). I doubt that this driver would improve in near future. Probably better to buy another device... (or as a very special solution for you, do the udev trick. But will you remember what you did, if something has to be fixed in 1 year?)

**Reply by LundSoftwares** [reply]
Thanks for the response
I'll try the fix by @mr-manuel first and then go for the udev, I already got a list of "todo" if reinstallation is needed so it will be added there if necessary :)

**Comment by mr-manuel** [comment]
If you want you can try my latest beta of this driver. It includes an option to exclude serial adapters to be recognized from serialbattery. Take a look at the `config.default.ini` of the last beta.

https://github.com/mr-manuel/venus-os_dbus-serialbattery/releases/tag/v1.0.20230605beta

https://github.com/mr-manuel/venus-os_dbus-serialbattery/blob/43cf469ba001656adc5e01c6f9da735296cd3b37/etc/dbus-serialbattery/config.default.ini#L194-L196

**Reply by LundSoftwares** [reply]
Great, I will try that next time I'm at the boat (or if I find a way to SSH into it from home).
I'll report back when tested!

**Reply by mr-manuel** [reply]
This could be interesting for you too https://github.com/mr-manuel/venus-os_ngrok

**Reply by LundSoftwares** [reply]
Awesome! 
Took two attempts to get the Token right, but now I can SSH straight in. Big Thanks!!

**Reply by LundSoftwares** [reply]
Oki Latest version installed now but no complete success. The "exclude port" seems to work, the SmartShunt popped right up now, but the Sinowealth BMS was not to be found. 

Used these settings: 
edit dbus-serialbattery.py to include Sinowealth:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/23386303/84e630b0-d219-4663-bec4-dd49c9f0843b)

config.ini (SmartShunt on USB1, BMS on USB2:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/23386303/e83af3a1-d19d-4e66-8b89-d7fa82d406ff)

Output from Log says that no BMS is available on USB2:
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/23386303/ee429e1f-884b-498f-80a6-0f628f77631a)

However I can confirm it's there and responding. 
Got an older Python script (https://github.com/dreadnought/python-daly-bms) that I can manually trigger from NodeRed and it still works.
![image](https://github.com/Louisvdw/dbus-serialbattery/assets/23386303/dba9854b-c57d-4ce5-8672-cec0df8243b0)

Any ideas, have I forgotten to do some settings? :(
(EDIT: Backed to v0.14.3 again and the BMS popped right in)

**Comment by mr-manuel** [comment]
Link to issue https://github.com/Louisvdw/dbus-serialbattery/issues/702
